Zend/zend_ast.c

```diff
@@ -441,6 +441,14 @@ ZEND_API int zend_ast_evaluate(zval *result, zend_ast *ast, zend_class_entry *sc
 				zend_error_noreturn(E_COMPILE_ERROR, "Cannot use [] for reading");
 			}
 
+			if (ast->child[0]->kind == ZEND_AST_DIM) {
+				if (ast->attr == ZEND_DIM_IS) {
+					ast->child[0]->attr = ZEND_DIM_IS;
+				} else {
+					ast->child[0]->attr = ZEND_DIM_PROPAGATE_ERROR;
+				}
+			}
```

Comment:
These checks should be located at https://github.com/php/php-src/blob/master/Zend/zend_compile.c#L8527 rather than here. Flags have to be set during compilation, because the AST cannot be safely modified at runtime.